### PR TITLE
Reusable Blocks: Add reusable blocks UI

### PR DESCRIPTION
**Pull request by noisysocks**
Adds the UI necessary for [reusable blocks](https://github.com/WordPress/gutenberg/issues/1516). This includes:

 - PHP code that renders a reusable block in a published post or page
 - The component rendered in the visual editor
 - The sidebar UI used to attach and detach reusable blocks
 - The inserter UI used to insert reusable blocks to the post

Some of this was already reviewed in https://github.com/WordPress/gutenberg/pull/2659.

### CODE REVIEW

File: editor/effects.js [hunk 1]
```diff
@@ -276,4 +288,78 @@ export default {
...
+		);
+	},
+	SAVE_REUSABLE_BLOCK( action, store ) {
+		const { id } = action;
+		const { getState, dispatch } = store;
+
+		const { name, type, attributes } = getReusableBlock( getState(), id );
+		const content = serialize( createBlock( type, attributes ) );
+
+		new wp.api.models.ReusableBlocks( { id, name, content } ).save().then(
```
Comment (lamosty) [inline editor/effects.js]: What is the `name` property? The WP REST API [doesn't support](https://developer.wordpress.org/rest-api/reference/posts/#create-a-post) a `name` attribute. Shouldn't it be `title` of the custom post type?

**Comment by karmatosed** [comment]
I wanted to test this to give some feedback, however I saw this:

2017-11-08 at 21 13

Is this a known issue or something I need to do to get it working?

**Comment by lamosty** [comment]
@karmatosed:
> I wanted to test this to give some feedback, however I saw this:

After you checkout the branch, try deleting `node_modules`, then `npm install` and `npm run build` again. I had the same issue.

**Comment by karmatosed** [comment]
I'm cc'ing in @jasmussen as I think we need to iterate on the design here. Currently we have this:

2017-11-10 at 20 45

This I am suggesting could be iterated by having some more friendly and informative language.

Here are some suggestions:

**Comment by jasmussen** [comment]
I think using the ellipsis menu for making reusable blocks has potential. I'm _slightly_ concerned about putting more stuff in there as we're already looking at a few options there, with a bunch of transformations to come also. But it does feel like a good place for it.

**Comment by karmatosed** [comment]
@jasmussen I have that concern myself. The current implementation feels like hiding a good feature though. Maybe we need to try it? Do we have a best guess at how many things we are going to have if everything we want gets in today? I'm thinking of a visual stress test.

**Comment by jasmussen** [comment]
I think this may be one of those features whose full potential won't be realized until slightly later on, perhaps post phase-1, where we have more templating features and so on. Given this, perhaps we shouldn't hyper-optimize for the current UI, but instead leave room for us to learn about the feature and then find a better place. To put it differently, there seems to be agreement that it's a slick feature that deserves a better place, but perhaps it's a problem that's best solved in the future?

**Comment by noisysocks** [comment]
I rebased this PR and fixed some bugs that were introduced in the last rebase! 🙂 

Since `block-inspector/advanced-controls.js` was removed in https://github.com/WordPress/gutenberg/pull/3472, I had to re-implement how the 'Convert to Reusable Block' control is added to the sidebar. It now uses a hook (`hooks/reusable-blocks.js`) which is very similar to how `hooks/custom-class-name.js` works. This does the job, but it feels a little hacky to me since it involves `connect()`ing the `InspectorControls`. 

The E2E tests are failing. A `git bisect` tells me that it's something in https://github.com/WordPress/gutenberg/pull/3378/commits/ce00056313e38c9329b583d215e47c0e5e40de78 that is causing them to break, but I can't for the life of me figure out what it is. My best guess is that Cypress doesn't like it when we use `Slot.Fill`? 😕

**Comment by karmatosed** [comment]
@jasmussen I still think it needs to move from where it is currently, I am ok if that for now is the ellipsis.

File: editor/effects.js (hunk 1 above)
Comment (noisysocks) [reply editor/effects.js]: `name` is the name of the reusable block: https://github.com/WordPress/gutenberg/blob/master/lib/class-wp-rest-reusable-blocks-controller.php#L302-L307

I agree that naming that `title` is a better thing to call that, though. I'll put it on my list!

Comment (lamosty) [reply editor/effects.js]: Ah, I see, thanks.

Your call. :)

**Comment by noisysocks** [comment]
Thanks for the 👀 @karmatosed! I'll move the controls into the ⋯ menu 🙂

**Comment by noisysocks** [comment]
> A git bisect tells me that something in 295e025 is causing them to break, but I can't for the life of me figure out what it is. My best guess is that Cypress doesn't like it when we use Slot.Fill? 😕

For posterity: turns out this was because `<Editable>` was being mounted twice as a result of the `BlockEdit` filter returning a deeply nested array rather than a flat array. 9472e0e was the fix. Thanks @youknowriad for the big hint in figuring this one out!

**Comment by noisysocks** [comment]
screen shot 2017-11-27 at 13 12 10

I moved the button into the block settings menu. @jasmussen—are you able to whip up an icon for this?

**Comment by jasmussen** [comment]
Nice work.

While the icon definitely could use some love, I think for now let's go with this dashicon https://developer.wordpress.org/resource/dashicons/#controls-repeat

Reusable blocks is an important interface that I expect we'll hone and polish many times in the future, but for now let's get the ball rolling!

**Comment by noisysocks** [comment]
Icon changed!

File: blocks/inspector-controls/button-control/index.js [hunk 2]
```diff
@@ -0,0 +1,23 @@
+/**
```
Comment (youknowriad) [inline blocks/inspector-controls/button-control/index.js]: This component is not used anymore, right?

File: editor/components/inserter/group.js [hunk 3]
```diff
@@ -48,7 +48,7 @@ export default class InserterGroup extends Component {
 				role="menuitem"
 				key={ block.name }
 				className="editor-inserter__block"
-				onClick={ selectBlock( block.name ) }
+				onClick={ selectBlock( block.name, block.attributes ) }
```
Comment (youknowriad) [inline editor/components/inserter/group.js]: Maybe we should just pass the `block` object as an argument here instead of two arguments.

File: editor/components/inserter/menu.js [hunk 4]
```diff
@@ -103,16 +107,16 @@ export class InserterMenu extends Component {
 		} );
 	}
 
-	selectBlock( name ) {
+	selectBlock( name, attributes ) {
 		return () => {
-			this.props.onSelect( name );
+			this.props.onSelect( name, attributes );
```
Comment (youknowriad) [inline editor/components/inserter/menu.js]: Thinking more about this "attributes" argument. Regular blocks already have an `attributes` property but instead of containing `attributes` values, it contains `attributes` definition. I assume this will be called with the `attributes` definitions for these regular blocks which is not what's expected I guess. 

It may not produce bugs at the moment but I think we should fix this ambiguity somehow.

File: blocks/inspector-controls/button-control/index.js (hunk 2 above)
Comment (noisysocks) [reply blocks/inspector-controls/button-control/index.js]: You're right! I've removed it.

File: editor/components/inserter/group.js (hunk 3 above)
Comment (noisysocks) [reply editor/components/inserter/group.js]: 👌 fixed in 8ced4510795d9e428a2f6d1c2bbb4e2f744da23b.

File: editor/components/inserter/menu.js (hunk 4 above)
Comment (noisysocks) [reply editor/components/inserter/menu.js]: 👌 fixed in 8ced4510795d9e428a2f6d1c2bbb4e2f744da23b.

**Comment by noisysocks** [comment]
🤔 looks like Travis CI is getting stuck on a `git clone`...

screen shot 2017-11-29 at 17 39 47
screen shot 2017-11-29 at 17 39 52

**Comment by mtias** [comment]
Looking good. A couple things.

I can search for "reusable block" and insert a broken block:

image

My saved block doesn't show up if I search for its name.

image

Let's add another tab to the inserter called "Saved" instead of putting saved blocks as a category of "blocks":

image

**Comment by mtias** [comment]
Another point is that the block inspector for the reusable block should let me delete the block from the records (with a warning that it'll stop working if it was used in a post):

image

**Comment by mtias** [comment]
Also for later: creation of global blocks should be handled with proper capabilities (an author would not get access to create reusable blocks, but an editor would, etc). And let this be filterable if someone wants to allow anyone to create reusable blocks.

File: blocks/library/reusable-block/index.php [hunk 5]
```diff
@@ -0,0 +1,44 @@
...
+	$block = array_shift( $blocks );
+	if ( ! $block ) {
+		return '';
+	}
+
+	return gutenberg_render_block( $block );
+}
+
+register_block_type( 'core/reusable-block', array(
+	'attributes'      => array(
```
Comment (mtias) [inline blocks/library/reusable-block/index.php]: Some odd spaces here.

File: blocks/library/reusable-block/index.php [hunk 6]
```diff
@@ -0,0 +1,44 @@
...
+
+	$block = array_shift( $blocks );
+	if ( ! $block ) {
+		return '';
+	}
+
+	return gutenberg_render_block( $block );
+}
+
+register_block_type( 'core/reusable-block', array(
```
Comment (mtias) [inline blocks/library/reusable-block/index.php]: All the other terms we can more easily change, like internal variables, actions, and functions. But this we'll have to support once we announce.

I am not super comfortable with the name. A couple ideas:

* `core/global-block`
* `core/saved-block` -> to match the inserter tab.
* or maybe this is an exception and is saved like `global-block` without the core suffix. Not to keen on an exception here, but it might warrant it (also of note, we have `core-embed` as a namespace.

For reference, the CPT is now called `wp-block`.

cc @aduth @youknowriad @mcsf @pento for naming thoughts

Comment (mtias) [reply blocks/library/reusable-block/index.php]: I'm not a native speaker, so if `reusable-block` is clearest I'm fine with sticking with it, just want to make sure we are ok with it.

File: editor/components/block-settings-menu/reusable-block-toggle.js [hunk 7]
```diff
@@ -0,0 +1,49 @@
...
+export function ReusableBlockToggle( { block, convertToStatic, convertToReusable } ) {
+	const isReusableBlock = block.name === 'core/reusable-block';
+
+	return (
+		<IconButton
+			className="editor-block-settings-menu__control"
+			icon="controls-repeat"
+			onClick={ isReusableBlock ? convertToStatic : convertToReusable }
+		>
+			{ isReusableBlock ? __( 'Detach from Reusable Block' ) : __( 'Make into a Reusable Block' ) }
```
Comment (mtias) [inline editor/components/block-settings-menu/reusable-block-toggle.js]: Would "Convert to Reusable Block" or "Convert to Global Block" be more concise?

File: lib/blocks.php [hunk 8]
```diff
@@ -131,6 +131,33 @@ function gutenberg_serialize_block( $block ) {
 	return $content;
 }
 
+/**
+ * Renders a single block into a HTML string.
+ *
+ * @since 1.3.0
```
Comment (mtias) [inline lib/blocks.php]: This needs to be updated.

File: lib/client-assets.php [hunk 9]
```diff
@@ -790,7 +790,7 @@ function gutenberg_editor_scripts_and_styles( $hook ) {
 	$script  = '( function() {';
 	$script .= sprintf( 'var editorSettings = %s;', wp_json_encode( $editor_settings ) );
 	$script .= <<<JS
-		window._wpLoadGutenbergEditor = Promise.all( [ wp.api.init(), wp.api.init( { versionString: 'gutenberg/v1' } ) ] ).then( function() {
+		window._wpLoadGutenbergEditor = Promise.all( [ wp.api.init(), wp.api.init( { versionString: 'gutenberg/v1/' } ) ] ).then( function() {
```
Comment (mtias) [inline lib/client-assets.php]: curious, what's the reason for this change?

File: blocks/library/reusable-block/index.php (hunk 6 above)
Comment (mtias) [reply blocks/library/reusable-block/index.php]: Thinking a bit more, it seems the most clear and consistent with the `wp-block` post type is to call this `core/block` which maps to `wp:block` and `wp:core/block`. Thoughts?

Comment (noisysocks) [reply blocks/library/reusable-block/index.php]: My only hesitation with _global block_ is that _global_ is a fairly technical term and that using it in the UI might limit how approachable this feature is to a wider audience. 

For technical usages though, I'm happy with using a more precise term. `core/block` matching up with `wp-block` is very appealing.

File: blocks/library/reusable-block/index.php (hunk 5 above)
Comment (noisysocks) [reply blocks/library/reusable-block/index.php]: The linter requires that the arrow on line 37 matches up with the arrow on line 43

File: lib/client-assets.php (hunk 9 above)
Comment (noisysocks) [reply lib/client-assets.php]: None of the Backbone methods work without the trailing slash. We didn't catch this while testing https://github.com/WordPress/gutenberg/pull/3377 since the unit tests mock the API call and no UI existed for any of this.

File: lib/blocks.php (hunk 8 above)
Comment (noisysocks) [reply lib/blocks.php]: Fixed in b8f96e9b5e7c905d37ba7e392083de1a249cc8a1.

File: editor/components/block-settings-menu/reusable-block-toggle.js (hunk 7 above)
Comment (noisysocks) [reply editor/components/block-settings-menu/reusable-block-toggle.js]: Changed in 619d81f51afbd3bf0fd5a407e2a82a993b7a7051.

File: blocks/library/reusable-block/index.js [hunk 10]
```diff
@@ -1,12 +1,160 @@
...
+
+		const reusableBlockAttributes = { ...reusableBlock.attributes, ...attributes };
+		const blockType = getBlockType( reusableBlock.type );
+		const BlockEdit = blockType.edit || blockType.save;
+
+		// Generate a class name for the block's editable form
+		const generatedClassName = hasBlockSupport( blockType, 'className', true ) ?
+			getBlockDefaultClassname( reusableBlock.type ) :
+			null;
+		const className = classnames( generatedClassName, reusableBlockAttributes.className );
```
Comment (noisysocks) [inline blocks/library/reusable-block/index.js]: Clever! Thanks so much @youknowriad 🙏

**Comment by noisysocks** [comment]
I rebased this and removed @youknowriad's Autocompleter changes since they were breaking tests. I'll add [those changes](https://gist.github.com/noisysocks/ddfe7d9e58f4e9bb59f5f33acaf1230a) back in a seperate PR.

I think that this is OK to merge. We can iterate on it in seperate (smaller) PRs. I've created some tickets for follow-up work: #3793, #3792, #3791

File: lib/client-assets.php (hunk 9 above)
Comment (youknowriad) [reply lib/client-assets.php]: Do you think this is still useful with https://github.com/WordPress/gutenberg/pull/3778

File: blocks/library/reusable-block/index.php (hunk 6 above)
Comment (mcsf) [reply blocks/library/reusable-block/index.php]: _Saved block_ might speak to users more. Further, though I'm not a huge fan, there are precedents out there to justify _custom blocks_. Note that I can see _reusable blocks_ working, if visual context and copy correctly introduce the concept.

Playing devil's advocate, _saved_ and _custom blocks_ may sound confusing ("Aren't I already making my own block when I add a "Paragraph" and type something in? Isn't it custom? Isn't it being saved?"). Those terms stress a different aspect from _reusable_, with _reusable_ arguably stressing the user-facing value. (Mind you, a similar case can be made against _reusable_: "This Paragraph, or this Author block, aren't they reusable?")

File: lib/client-assets.php (hunk 9 above)
Comment (noisysocks) [reply lib/client-assets.php]: Yes—we're not using `withAPIData`.

File: blocks/library/reusable-block/index.php (hunk 6 above)
Comment (noisysocks) [reply blocks/library/reusable-block/index.php]: I'm going to go ahead and rename the block identifier to `core/block` so that we can move ahead with this PR. This makes sense to me since, just as `<!-- wp:image -->` inserts an image, `<!-- wp:block -->` inserts a `wp_block`.

Let's move the discussion about the name of the overall feature to a seperate issue—for what it's worth, I asked my wife (a regular WordPress user) what she thought and she said that _custom block_ makes the most sense 🙂

**Comment by noisysocks** [comment]
Any last words before I pull the trigger on this?

File: blocks/library/block/edit-panel/index.js [hunk 11]
```diff
@@ -0,0 +1,68 @@
...
+					type="text"
+					disabled={ isSaving }
+					className="reusable-block-edit-panel__name"
+					value={ name }
+					onChange={ ( event ) => onChangeName( event.target.value ) } />,
+				<Button
+					key="save"
+					isPrimary
+					isLarge
+					isIndicatingProgress={ isSaving }
```
Comment (aduth) [inline blocks/library/block/edit-panel/index.js]: This prop does not exist on the `Button` component, and since Button passes through unrecognized props to the underlying DOM node, it results in a warning while editing:

>Warning: React does not recognize the `isIndicatingProgress` prop on a DOM element. If you intentionally want it to appear in the DOM as a custom attribute, spell it as lowercase `isindicatingprogress` instead. If you accidentally passed it from a parent component, remove it from the DOM element.

Should this be the `isBusy` prop added in #3682?

File: blocks/library/block/edit-panel/index.js [hunk 12]
```diff
@@ -0,0 +1,68 @@
...
+				</Button>,
+				<Button
+					key="detach"
+					isLarge
+					className="reusable-block-edit-panel__button"
+					onClick={ onDetach }>
+					{ __( 'Detach' ) }
+				</Button>,
+			] }
+			{ ( isEditing || isSaving ) && [
```
Comment (aduth) [inline blocks/library/block/edit-panel/index.js]: Should the editing variation be visible while the name is being saved? In my testing, it flips back to the static text immediately upon hitting Save.

File: blocks/library/block/index.js [hunk 13]
```diff
@@ -0,0 +1,172 @@
...
+					onChangeName={ this.setName }
+					onSave={ this.updateReusableBlock }
+					onCancel={ this.stopEditing }
+				/>
+			),
+		];
+	}
+}
+
+const ConnectedReusableBlockEdit = connect(
```
Comment (aduth) [inline blocks/library/block/index.js]: I don't have any ideas at the moment and it may have already been mentioned, but we need to work to find a solution where the block can access state without the implicit dependency on editor, either:

- Blocks module having its own separate state
- Merging more of [the non-post-editor-specific-] editor and blocks

Obvious too by the fact that we have equivalent action creators for what we're dispatching here, but unable (unwilling) to import them directly from editor.

File: blocks/library/block/index.js [hunk 14]
```diff
@@ -0,0 +1,172 @@
...
+		},
+	} )
+)( ReusableBlockEdit );
+
+registerBlockType( 'core/block', {
+	title: __( 'Reusable Block' ),
+	category: 'reusable-blocks',
+	isPrivate: true,
+
+	attributes: {
```
Comment (aduth) [inline blocks/library/block/index.js]: I don't think you should need to define this on the client, since it should be bootstrapped from the server definition.

See #2529

File: components/spinner/index.js [hunk 15]
```diff
@@ -1,3 +1,10 @@
-const spinner = <span className="spinner is-active" />;
+/**
+ * External dependencies
+ */
+import classnames from 'classnames';
 
-export default () => spinner;
+function Spinner( { className } ) {
```
Comment (aduth) [inline components/spinner/index.js]: You're not / no longer taking advantage of these enhancements? Should we omit them respecting our good friend [YAGNI](https://en.wikipedia.org/wiki/You_aren%27t_gonna_need_it)?

File: editor/components/block-list/block.js [hunk 16]
```diff
@@ -350,11 +350,13 @@ class BlockListBlock extends Component {
 		const { isHovered, isSelected, isMultiSelected, isFirstMultiSelected, focus } = this.props;
 		const showUI = isSelected && ( ! this.props.isTyping || ( focus && focus.collapsed === false ) );
 		const { error } = this.state;
+		const isReusableBlock = blockType.name === 'core/block';
```
Comment (aduth) [inline editor/components/block-list/block.js]: Elsewhere we've limited our awareness to specific block types via abstraction, e.g. `getDefaultBlockName`, `getUnknownTypeHandlerName`. Should we need to consider that here? Maybe not the case we'd ever expect it to change.

File: editor/components/block-settings-menu/reusable-block-toggle.js [hunk 17]
```diff
@@ -0,0 +1,49 @@
...
+import { IconButton } from '@wordpress/components';
+
+/**
+ * Internal dependencies
+ */
+import { getBlock } from '../../selectors';
+import { convertBlockToStatic, convertBlockToReusable } from '../../actions';
+
+export function ReusableBlockToggle( { block, convertToStatic, convertToReusable } ) {
+	const isReusableBlock = block.name === 'core/block';
```
Comment (aduth) [inline editor/components/block-settings-menu/reusable-block-toggle.js]: Thinking purely from the [presentational / container distinction](https://medium.com/@dan_abramov/smart-and-dumb-components-7ca2f9a7c7d0), I might expect this variable to be passed as the prop, determined in a selector or `mapStateToProps`, particularly since we're not otherwise using the `block` prop.

File: editor/components/inserter/menu.js [hunk 18]
```diff
@@ -71,6 +71,10 @@ export class InserterMenu extends Component {
 		this.switchTab = this.switchTab.bind( this );
 	}
 
+	componentDidMount() {
+		this.props.fetchReusableBlocks();
```
Comment (aduth) [inline editor/components/inserter/menu.js]: Noting that this can become quite noisy for AJAX requests: Every time the user opens an inserter.

Personally I'm okay with this, since it ensures that the data is kept fresh, but worth keeping in mind.

**Review by aduth — COMMENTED** [review]
Most comments aren't critical. This works great and is _very_ cool to use and see 👍

File: blocks/library/block/edit-panel/index.js (hunk 12 above)
Comment (noisysocks) [reply blocks/library/block/edit-panel/index.js]: Good catch—that's a bug that must have slipped in during a merge. Fixed in a820d1cc43dddd97d44e6e756707dfc6ab56c80b.

File: blocks/library/block/edit-panel/index.js (hunk 11 above)
Comment (noisysocks) [reply blocks/library/block/edit-panel/index.js]: Good catch. It definitely should be `isBusy`. Fixed in a820d1cc43dddd97d44e6e756707dfc6ab56c80b.

File: blocks/library/block/index.js (hunk 13 above)
Comment (noisysocks) [reply blocks/library/block/index.js]: Agreed, and I also have no ideas on how to address this right now. 

> Merging more of [the non-post-editor-specific-] editor and blocks

Could you elaborate on this? I'm not quite caught up on my history here. Did we abandon the effort to merge the two modules (https://github.com/WordPress/gutenberg/pull/2795)?

File: blocks/library/block/index.js (hunk 14 above)
Comment (noisysocks) [reply blocks/library/block/index.js]: I'm hesitant to do this because removing the client-side definition means our test fixtures aren't able to test parsing a `<!-- wp:block ref="358b59ee-bab3-4d6f-8445-e8c6971a5605" /-->`.

File: components/spinner/index.js (hunk 15 above)
Comment (noisysocks) [reply components/spinner/index.js]: 👍 I've removed these changes.

File: editor/components/block-list/block.js (hunk 16 above)
Comment (noisysocks) [reply editor/components/block-list/block.js]: Couldn't hurt! I moved this logic out to `isReusableBlock` in 95a28a37adea3afeb44fb2b9c068c9e1fc61fb31.

File: editor/components/block-settings-menu/reusable-block-toggle.js (hunk 17 above)
Comment (noisysocks) [reply editor/components/block-settings-menu/reusable-block-toggle.js]: 👍 I've moved it in 95a28a37adea3afeb44fb2b9c068c9e1fc61fb31.

File: editor/components/inserter/menu.js (hunk 18 above)
Comment (noisysocks) [reply editor/components/inserter/menu.js]: For now I think this is OK. If it's a problem then we can limit it to fetch once when the inserter is opened for the first time.

File: blocks/library/block/index.js (hunk 13 above)
Comment (youknowriad) [reply blocks/library/block/index.js]: Just an idea: Does it make more sense for this block to be defined in the `editor` module instead?

I also even wonder if the `blocks` module should hold only the blocks API and define all the blocks in the `editor` module

**Comment by mtias** [comment]
@noisysocks Before we go ahead with exposing this at wide. Can you remind me why we are not using the ID provided by WP for the post object in `wp_block` like `public 'ID' => int 2359` instead of the generated `ref`? It makes a lot more sense for me to see it as `<!-- wp:block id="2359" /-->` which is an internally consistent concept of WordPress.

File: blocks/library/block/index.js (hunk 13 above)
Comment (mcsf) [reply blocks/library/block/index.js]: > Just an idea: Does it make more sense for this block to be defined in the `editor` module instead?

I think there's a lot of merit to that idea.

> I also even wonder if the `blocks` module should hold only the blocks API and define all the blocks in the `editor` module

That one I'm not too convinced of right now. I mean, I could picture using `editor/library`, but the current situation of having the library sit next to the API makes sense to me, as the core blocks are the canonical blocks: in that sense, they highly complement the API (both code and docs) by providing numerous well tested use cases.

File: editor/reducer.js [hunk 19]
```diff
@@ -731,6 +754,22 @@ export const reusableBlocks = combineReducers( {
...
+				const { id, updatedId } = action;
+
+				// If a temporary reusable block is saved, we swap the temporary id with the final one
+				if ( id === updatedId ) {
+					return state;
+				}
+				return {
+					...omit( state, id ),
+					[ updatedId ]: {
+						...omit( state[ id ], [ 'id', 'isTemporary' ] ),
```
Comment (mcsf) [inline editor/reducer.js]: Totally minor, but omitting `'id'` isn't strictly necessary.

**Review by mcsf — APPROVED** [review]
Per Slack, this looks great. The recent change to server-issued IDs is quite welcome.

**Comment by mtias** [comment]
Great work here everyone and specially @noisysocks. This is a great feature.

**Comment by noisysocks** [comment]
> Also for later: creation of global blocks should be handled with proper capabilities (an author would not get access to create reusable blocks, but an editor would, etc). And let this be filterable if someone wants to allow anyone to create reusable blocks.

@mtias — could you please elaborate on this so that I can create an issue? Right now, any user with the `edit_posts` capability (authors, contributors and editors) can view and edit Reusable Blocks. Are you suggesting that we limit editing and creating Reusable Blocks to _just_ editors?

**Comment by mtias** [comment]
> Are you suggesting that we limit editing and creating Reusable Blocks to just editors? Right now, any user with the edit_posts capability (authors, contributors and editors) can view and edit Reusable Blocks.

We need to let authors insert them, but probably not edit them by default. Not sure if we should let authors create them, or only editors. Furthermore, this is something that should probably be filterable and part of general capabilities so people can map. Example: `edit_blocks`, `insert_blocks`, `delete_blocks`. Something more broad to think about. cc @pento

**Comment by tkes** [comment]
@mtias ... something more broad.

There indeed needs to be a complete capabilities- and author-concept for reusable blocks.

A current users with a role below "editor" should only be able to see/create/edit/delete **his/her own** reusable blocks.

Users with/above "editor" role should be able to see/create/edit/delete **any** reusable block and also edit its author.

The concept should also provide a way to configure the visibility of reusable blocks for each user and/or post_type, not sure if via capability or filter.

Example for no visibility limit requirement: Simple one-user blog.

Example for visibility limit requirement: Multi author website with 15-20 authors and everybody creates a few reusable blocks which would then be visible for everybody, the gui would be a mess in no time...
